internal/cephfs/groupcontrollerserver.go

```diff
@@ -0,0 +1,758 @@
+	volIDs []string,
+	cr *util.Credentials) (
+	map[string]core.FSQuiesceClient,
+	error,
+) {
+	type fs struct {
+		fsName                string
+		volumes               []core.Volume
+		subVolumeGroupMapping map[string][]string
+		monitors              string
+	}
+	fm := make(map[string]fs, 0)
+	for _, volID := range volIDs {
+		// Find the volume using the provided VolumeID
+		volOptions, _, err := store.NewVolumeOptionsFromVolID(ctx,
+			volID, nil, secret, "", false)
+		if err != nil {
+			return nil, err
+		}
+		volOptions.Destroy()
+		// choosing monitorIP's and fsName as the unique key
+		// TODO: Need to use something else as the unique key as users can
+		// still choose the different monitorIP's and fsName for subvolumes
+		uniqueName := volOptions.Monitors + volOptions.FsName
+		if _, ok := fm[uniqueName]; !ok {
+			fm[uniqueName] = fs{
+				fsName:                volOptions.FsName,
+				volumes:               make([]core.Volume, 0),
+				subVolumeGroupMapping: make(map[string][]string), // Initialize the map
+				monitors:              volOptions.Monitors,
+			}
+		}
+		a := core.Volume{
+			VolumeID:  volID,
+			ClusterID: volOptions.ClusterID,
+		}
+		// Retrieve the value, modify it, and assign it back
+		val := fm[uniqueName]
+		val.volumes = append(val.volumes, a)
+		existingVolIDInMap := val.subVolumeGroupMapping[volOptions.SubVolume.SubvolumeGroup]
+		val.subVolumeGroupMapping[volOptions.SubVolume.SubvolumeGroup] = append(
+			existingVolIDInMap,
+			volOptions.SubVolume.VolID)
+		fm[uniqueName] = val
+	}
+	fsk := map[string]core.FSQuiesceClient{}
+	var err error
+	for k, v := range fm {
+		conn := &util.ClusterConnection{}
+		if err = conn.Connect(v.monitors, cr); err != nil {
+			return nil, err
+		}
+		fsk[k], err = core.NewFSQuiesce(v.fsName, v.volumes, v.subVolumeGroupMapping, conn)
+		if err != nil {
+			log.ErrorLog(ctx, "failed to get subvolume quiesce: %v", err)
+
+			return nil, err
+		}
+	}
+
+	return fsk, nil
+}
+
+// matchesSourceVolumeIDs checks if the sourceVolumeIDs and volumeIDsInOMap are
+// equal.
+func matchesSourceVolumeIDs(sourceVolumeIDs, volumeIDsInOMap []string) bool {
+	// sort the array as its required for slices.Equal call.
+	sort.Strings(sourceVolumeIDs)
+	sort.Strings(volumeIDsInOMap)
+
+	return slices.Equal(sourceVolumeIDs, volumeIDsInOMap)
+}
+
+// deleteSnapshotsAndUndoReservation deletes the snapshots and undoes the
+// volume group reservation. It also resets the quiesce of the subvolumes and
+// subvolume groups in the filesystems for the volumeID's present in the
+// CreateVolumeGroupSnapshotRequest.
+func (cs *ControllerServer) deleteSnapshotsAndUndoReservation(ctx context.Context,
+	vgs *store.VolumeGroupSnapshotIdentifier,
+	cr *util.Credentials,
+	fsMap map[string]core.FSQuiesceClient,
+	secrets map[string]string,
+) error {
+	// get the omap from the snapshot and volume mapping
+	bgi, vgsi, err := store.NewVolumeGroupOptionsFromID(ctx, vgs.VolumeGroupSnapshotID, cr)
```

Comment:
`NewVolumeGroupOptionsFromID()` returns a connected `VolumeGroupOptions`. A `defer bgi.Destroy()` seems to be missing.

Reply:
yes its a good catch, fixed it now